src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java

```diff
@@ -3515,42 +3547,54 @@ JCVariableDecl variableDeclarator(JCModifiers mods, JCExpression type, boolean r
+                log.error(DiagnosticFlag.SYNTAX, pos, Errors.UnderscoreAsIdentifier);
+            }
+        }
+
         if (token.kind == EQ) {
             nextToken();
             init = variableInitializer();
         }
         else if (reqInit) syntaxError(token.pos, Errors.Expected(EQ));
-        JCTree elemType = TreeInfo.innermostType(type, true);
-        int startPos = Position.NOPOS;
-        if (elemType.hasTag(IDENT)) {
-            Name typeName = ((JCIdent)elemType).name;
-            if (restrictedTypeNameStartingAtSource(typeName, pos, !compound && localDecl) != null) {
-                if (typeName != names.var) {
-                    reportSyntaxError(elemType.pos, Errors.RestrictedTypeNotAllowedHere(typeName));
-                } else if (type.hasTag(TYPEARRAY) && !compound) {
-                    //error - 'var' and arrays
-                    reportSyntaxError(elemType.pos, Errors.RestrictedTypeNotAllowedArray(typeName));
-                } else {
-                    declaredUsingVar = true;
-                    if(compound)
-                        //error - 'var' in compound local var decl
-                        reportSyntaxError(elemType.pos, Errors.RestrictedTypeNotAllowedCompound(typeName));
-                    startPos = TreeInfo.getStartPos(mods);
-                    if (startPos == Position.NOPOS)
-                        startPos = TreeInfo.getStartPos(type);
-                    //implicit type
-                    type = null;
+        JCVariableDecl result;
+        if (!isTypePattern) {
+            int startPos = Position.NOPOS;
+            JCTree elemType = TreeInfo.innermostType(type, true);
+            if (elemType.hasTag(IDENT)) {
+                Name typeName = ((JCIdent) elemType).name;
+                if (restrictedTypeNameStartingAtSource(typeName, pos, !compound && localDecl) != null) {
+                    if (typeName != names.var) {
+                        reportSyntaxError(elemType.pos, Errors.RestrictedTypeNotAllowedHere(typeName));
+                    } else if (type.hasTag(TYPEARRAY) && !compound) {
+                        //error - 'var' and arrays
+                        reportSyntaxError(elemType.pos, Errors.RestrictedTypeNotAllowedArray(typeName));
+                    } else {
+                        declaredUsingVar = true;
+                        if (compound)
+                            //error - 'var' in compound local var decl
+                            reportSyntaxError(elemType.pos, Errors.RestrictedTypeNotAllowedCompound(typeName));
+                        startPos = TreeInfo.getStartPos(mods);
+                        if (startPos == Position.NOPOS)
+                            startPos = TreeInfo.getStartPos(type);
+                        //implicit type
+                        type = null;
+                    }
                 }
             }
+            result = toP(F.at(pos).VarDef(mods, name, type, init, declaredUsingVar));
```

Comment:
I think this should check whether name is underscore (and feature is enabled) and, if that's the case, replace the name with an empty name, so that the rest of the compiler is already correct.

Also, related, no matter what, I think javac's Pretty class should be updated to render variables with empty name as "_" when the tree is turned back into source.

Reply:
yep I agree with this, the parser should set the right name as soon as it sees a `_`

Reply:
We could indeed if we wanted to store `empty` everywhere. However, after some off-list discussions I updated the CSR (and the code) to include `empty` as the name at the symbol-level (for consistency with javax model) and `_` at the Tree API-level for its version of a name. This is the rationale behind what you see in Attr (for patterns) and member enter for all other cases. @jddarcy what do you think?

Reply:
> 

In terms of pretty printing, src/jdk.compiler/share/classes/com/sun/tools/javac/processing/PrintingProcessor.java may need to be updated too.

Reply:
> 

For the user-facing javax.lang.model languages, I think the current convention of "empty name means unnamed" is reasonable and is consistent with how other unnamed items are handled in that API.

I assumed for the lower-level tree API "_" would be a more appropriate choice, but I'm open to discussions alternative conventions there.

Reply:
Recently, I realized that `names.empty` also signifies an error in `variableDeclaratorId` in some occasion around lambdas:  https://github.com/openjdk/jdk/blob/ffae4e0e97b19b1c1ad2865ed618ba380a861b7e/src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java#L3688-L3697
I don't believe that this would be a blocker but it could pose a possible minor challenge.

Reply:
FWIW, I don't have a strong opinion on whether the name in the AST should be empty or underscore. (But I agree empty name is more consistent in the Language Model.) For missing names where names are expected (which is the above case), we could revisit later, there are more cases like this.